chia/data_layer/data_store.py

```diff
@@ -234,18 +237,47 @@ async def insert_into_data_store_from_file(
                 if node_hash not in internal_nodes and node_hash not in terminal_nodes:
                     missing_hashes.append(node_hash)
 
+        root = await self.get_tree_root(store_id=store_id)
+        latest_blob = await self.get_merkle_blob(root.node_hash, read_only=True)
+        known_hashes = {}
+        if not latest_blob.empty():
+            nodes = latest_blob.get_nodes_with_indexes()
+            known_hashes = {node.hash: index for index, node in nodes}
+
+        new_missing_hashes: list[bytes32] = []
+        for hash in missing_hashes:
+            if hash in known_hashes:
+                assert root.node_hash is not None
+                merkle_blob_queries[root.node_hash].append(known_hashes[hash])
+            else:
+                new_missing_hashes.append(hash)
+
+        missing_hashes = new_missing_hashes
+        if missing_hashes:
+            async with self.db_wrapper.reader() as reader:
+                cursor = await reader.execute(
+                    "SELECT MAX(generation) FROM nodes WHERE store_id = ?",
+                    (store_id,),
+                )
+                row = await cursor.fetchone()
+                if row is None or row[0] is None:
+                    current_generation = 0
+                else:
+                    current_generation = row[0]
+
         batch_size = min(500, SQLITE_MAX_VARIABLE_NUMBER - 10)
-        found_hashes: set[bytes32] = set()
 
-        async with self.db_wrapper.reader() as reader:
-            for batch in to_batches(missing_hashes, batch_size):
-                placeholders = ",".join(["?"] * len(batch.entries))
-                query = f"""
-                    SELECT hash, root_hash, idx
-                    FROM nodes
-                    WHERE store_id = ? AND hash IN ({placeholders})
-                    LIMIT {len(placeholders)}
-                """
+        while missing_hashes:
+            found_hashes: set[bytes32] = set()
+            async with self.db_wrapper.reader() as reader:
+                for batch in to_batches(missing_hashes, batch_size):
+                    placeholders = ",".join(["?"] * len(batch.entries))
+                    query = f"""
+                        SELECT hash, root_hash, idx
+                        FROM nodes
+                        WHERE store_id = ? AND hash IN ({placeholders})
+                        LIMIT {len(placeholders)}
+                    """
 
                 async with reader.execute(query, (store_id, *batch.entries)) as cursor:
```

Comment:
Needs another indent now to stay inside the for loop?

Reply:
yes, thanks for catching this